### ISSUE TITLE
[API] ItemMeta DestroyableKeys/PlaceableKeys methods are disappeared in Paper API 1.20.5+

### ISSUE THREAD

**Issue by Dymeth** [PaperSpigot/Paper#10818]
### Expected behavior

**The problem:**

The backward capability of there's methods was broken in Paper 1.20.5. No more there's methods exist on Paper API 1.20.5+:
```
ItemMeta#getDestroyableKeys()
ItemMeta#getPlaceableKeys()
ItemMeta#hasDestroyableKeys()
ItemMeta#hasPlaceableKeys()
ItemMeta#setDestroyableKeys(@NotNull Collection<Namespaced> canDestroy)
ItemMeta#setPlaceableKeys(@NotNull Collection<Namespaced> canPlaceOn)
```

Comparation:
1.20.4 API: https://jd.papermc.io/paper/1.20.4/org/bukkit/inventory/meta/ItemMeta.html
1.20.5 API: https://jd.papermc.io/paper/1.20.5/org/bukkit/inventory/meta/ItemMeta.html

Ways to solve i see:
- Reimplement listed methods and mark it as deprecated if necessary.
- If it hard to implement correctly, it must just do nothing (with print warning to logs most likely).
- It can also throw exception with understandable description (not preferred as it will break plugins logic).

### Observed/Actual behavior

Paper 1.20.6 marked as stable for now, but still there's no methods.
Methods wasn't marked as deprecated and alterative wasn't suggested.
Now it looks like the methods just magically disappeared or never existed at all.

There's methods was removed too, but it was marked as deprecated earlier:
```
ItemMeta#getCanDestroy()
ItemMeta#getCanPlaceOn()
ItemMeta#setCanDestroy(Set<Material> canDestroy)
ItemMeta#setCanPlaceOn(Set<Material> canPlaceOn)
ItemMeta#setDisplayNameComponent(@Nullable net.md_5.bungee.api.chat.BaseComponent[] component)
```

### Steps/models to reproduce

Just try to use listed methods in your plugin with Paper API 1.20.5 or newer

### Plugin and Datapack List

-

### Paper version

Paper version 1.20.6-115-master@9d6f2cc (2024-05-28T14:55:16Z) (Implementing API version 1.20.6-R0.1-SNAPSHOT)

### Other

Also possibly related with #8128

**Comment by electronicboy** [comment]
Yes, this was noted in the release notes. That API was no longer able to represent the underlying data, and keeping it around as-is would just break itemstacks, basically

**Comment by lynxplay** [comment]
We are 100% aware of it, but it is a terribly difficult thing to expose, given that the mechanic now offers a way to match the nbt of block entities.

The API does not have such a concept at all, so we'll need to find some ugly middleground to get something like this in, which is never fun and just a lot of back and forth.

**Comment by lynxplay** [comment]
Sorry for the long delay on this implementation, we went through a lot of back and forth on what the best way forward here is.
We ended up bringing back the old API methods, now deprecated for removal.

As a bit of input as to why, these methods no longer represent the entire can_place and can_destroy components of an item.
As such, both getters and setters return/set incomplete values, which will mean that using these methods may remove data from items generated from other sources.
This is obviously less than idea, but these methods will stay around as a bandage fix until we manage to provide a correct alternative.

Sorry again for the long wait, use the methods with caution ^
